src/pudl/transform/ferc1.py

```diff
@@ -6195,26 +6316,67 @@ def calculation_components_xbrl_ferc1(**kwargs) -> pd.DataFrame:
 
+    # Defensive testing on this table!
     assert calc_components[["table_name", "xbrl_factoid"]].notnull().all(axis=1).all()
 
-    calc_cols = ["table_name", "xbrl_factoid"] + dimensions
-    calc_and_parent_cols = calc_cols + [f"{col}_parent" for col in calc_cols]
+    # Let's check that all calculated components that show up in our data are
+    # getting calculated.
+    def check_calcs_vs_table(
+        calcs: pd.DataFrame,
+        checked_table: pd.DataFrame,
+        idx_calcs: list[str],
+        idx_table: list[str],
+        how: Literal["in", "not_in"],
+    ) -> pd.DataFrame:
+        if how == "in":
+            idx = calcs.set_index(idx_calcs).index.intersection(
+                checked_table.set_index(idx_table).index
+            )
+        elif how == "not_in":
+            idx = calcs.set_index(idx_calcs).index.difference(
+                checked_table.set_index(idx_table).index
+            )
+        calcs_vs_table = calcs.set_index(idx_calcs).loc[idx]
+        return calcs_vs_table.reset_index()
 
-    missing_from_calcs_idx = (
-        calc_components[calc_components.table_name.isin(FERC1_TFR_CLASSES.keys())]
-        .set_index(calc_cols)
-        .index.difference(metadata_xbrl_ferc1.set_index(calc_cols).index)
+    # which calculations are missing from the metadata table?
+    missing_calcs = check_calcs_vs_table(
+        calcs=calc_components[
+            calc_components.table_name.isin(FERC1_TFR_CLASSES.keys())
+        ],
+        checked_table=metadata_xbrl_ferc1,
+        idx_calcs=calc_cols,
+        idx_table=calc_cols,
+        how="not_in",
     )
     # ensure that none of the calculation components that are missing from the metadata
     # table are from any of the exploded tables.
-    missing_calcs = calc_components.set_index(calc_cols).loc[missing_from_calcs_idx]
     if not missing_calcs.empty:
-        raise AssertionError(
-            # logger.warning(
-            f"Found missing calculations from the exploded tables:\n{missing_calcs=}"
+        logger.warning(
+            "Calculations found in calculation components table are missing from the "
+            "metadata_xbrl_ferc1 table."
+        )
+        # which of these missing calculations actually show up in the transformed tables?
+        # Ignore any 'missing' calc components that aren't in the actual transformed tables.
+        actually_missing_kids = check_calcs_vs_table(
+            calcs=missing_calcs,
+            checked_table=table_dimensions_ferc1,
+            idx_calcs=child_cols,
+            idx_table=child_cols,
+            how="in",
+        )
+        logger.warning(
+            f"{len(actually_missing_kids)} of {len(missing_calcs)} missing calculation components observed in transformed FERC1 data."
         )
+        if not actually_missing_kids.empty:
+            raise AssertionError(
+                f"Found missing calculations from the exploded tables:\n{actually_missing_kids=}"
+            )
+
```

Comment:
Not all child components show up in all years of data, but because of the way we're editing calcs to account for the older DBF-only factoids, they'll always be added to the metadata tables. When we run the fast ETL there's an issue where these DBF-only factoids, which exist in our metadata but not in our actual tables, are getting flagged as missing calculations (the parent component does actually exist, the child does not).

Commit in which change was made: https://github.com/catalyst-cooperative/pudl/commit/913e5fbba2bb472e9a3159dfc371563cf477d559

You can see these failures here: https://github.com/catalyst-cooperative/pudl/actions/runs/6344230312/job/17233881055

Reply:
I added some more description to this code to clarify this, however!